versioning/worker-versions.md

```diff
@@ -0,0 +1,174 @@
+  attempted by new workers, and if they succeed, that workflow will subsequently
+  only be processed by new workers. If they fail, execution goes back to old
+  workers.
+
+# Using opaque whole-worker-identifiers as worker versions
+
+Thanks to @jlegrone for his great feedback encouraging us to revive this approach.
+
+We already have a field called `binaryChecksum` (or similar) in `WorkerOptions`
+today in SDKs. This field's intent is to uniquely identify a specific worker
+build. Since we believe versioning entire workers is the easiest solution, we
+can lean on this existing field.
+
+Deprecating the name `binaryChecksum`and replacing it with something more 
+accurate like `workerBuildId` would be appropriate, since most of our SDKs
+don't even produce a binary.
+
+With this approach, users who wish to make use of the improved versioning 
+features should set the `workerBuildId` during their CI process to a value that
+is unique for that worker's implementation. _Implementation_ rather than 
+_build_ is an important distinction here, since rebuilds of identical
+code should (ideally) produce workers wit the same ID. It is not necessary to
+do this, and it is varyingly difficult depending on the language, but when
+done it enables users to produce workers with the same ID from their CI builds
+if they didn't touch worker code, but something else (ex: The CI definition itself).
+This involves some work on the part of the user, but is more reliable than expecting
+worker authors to increment a version number.
+
+Unfortunately it is infeasible to automatically set an appropriate ID for many languages.
+
+## How is versioning accomplished with the `workerBuildId`?
+
+Every new worker ID will be assumed to be incompatible with the existing id(s)
+unless explicitly specified otherwise. New workers will not process *anything* 
+until explicitly enabled if they opt into build-id based versioning. Users who 
+wish to always have new workers begin processing new workflows can make an API 
+call immediately establishing the new worker as latest as part of their 
+deployment process (see below).
+
+In order to either (safely or not) migrate *open* workflows to a new version,
+users will explicitly indicate that an ID is compatible with some already-extant
+ID.
+
+Because this behavior is a breaking change, SDKs will need to add a
+`enableBuildIdBasedVersioning` boolean flag to their `WorkerOptions`.
+
+## Server-side APIs for `workerBuildId`-based versioning
+
+Importantly, we _already_ send the id (checksum, today) on every poll request.
+It's TBD if the worker will be changed to poll a unique queue per ID, or if
+that will happen on the frontend, or matching.
+
+These APIs could very well end up in a different service, but are shown
+as part of `WorkflowService` here for simplicity.
+```protobuf
+service WorkflowService {
+    rpc SetWorkerBuildIdOrdering (SetWorkerBuildIdOrderingRequest) returns (SetWorkerBuildIdOrderingResponse) {}
```

Comment:
Can this only be run once per worker build ID or is it expected to be able to be repeatedly called to make changes? Can I unset latest or change compatibility? Or should it error if I try to change anything (but accept same-parameters since this may be called on each worker deploy)?

Reply:
You can call it more than once and it'll reflect the latest write.

Reply:
So if I call it with `latest` unset after calling with `latest=true`, does that mean there is no latest anymore?

Reply:
No -- it'll be an error since you didn't change anything. Per https://github.com/temporalio/proposals/pull/54#discussion_r830183599 I think we would want to be very explicit about "unsetting" latest by setting it to the previous (or whatever other) version.

Reply:
There's no way to distinguish between you setting it false explicitly, and it being unset as a protobuf, so like most protobuf bools it should only have an effect when set true

Reply:
There are many ways to handle this. You could use `google.protobuf.BoolValue` or you could change the name to `move_to_latest` or anything of the sort.

I don't think it matters that much of course. It's just the concern that happens when sharing a always-important readable value with an only-sometimes-important writable value.